Cleanup climate and ecobee

Duoxilian opened the pull request:
**Description:**

When creating hold_mode, a number of cleanup opportunities to __init__.py and ecobee.py were observed. This pull request performs these.

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

Duoxilian committed:
Remove redundant input validation which is already accomplished through
defined schemata.

Duoxilian committed:
Rely on defined state attributes for hold mode.

homeassistant commented:
Hi @Duoxilian,

It seems you haven't yet signed a CLA. Please do so [here](https://cla.home-assistant.io/sign/Duoxilian?pr=5616).

Once you do that we will be able to review and accept this pull request.

Thanks!

Duoxilian committed:
Remove misleading comment. This comment seems to assume that sleep
mode is a hold; it is a schedule instead. The code snippets in the
comment could never work.

Duoxilian commented:
The request to sign the CLA results in this error:
```{"message":"Missing Authentication Token"}```

balloob commented:
CC @robbiet480

MartinHjelmare commented on homeassistant/components/climate/__init__.py:
```diff
@@ -43,6 +43,9 @@
 STATE_FAN_ONLY = "fan_only"
+STATE_AWAY_HOLD = "away"
+STATE_HOME_HOLD = "home"
+STATE_TEMPERATURE_HOLD = "temp"
```
Are these really states? At least for some of them it seems the device will still be in heat or cold state at the same time as using one of these modes. They are meta states sort of.

nordlead2005 replied:
In terms of the ecobee, "home", and "away" are "climates", which define various parameters on the thermostat. I don't know what they mean on other thermostats.

I also don't know if it makes sense to define strings that are currently only being used by one thermostat (ecobee) in __init__. Will they be re-used in other thermostats?

Duoxilian replied:
@nordlead2005 Actually, in ecobee terminology, "home" and "away" are **schedules**. In hass they hey are the state of the mode variable. Then there is "smart home" and "smart away" (and others) which are "hold states". The latter is this what the constants above refer to. "Away" is already used by a number of other thermostats. I would think the others would end up being implemented also in others that support different hold modes. 

@MartinHjelmare I guess that depends on what you think of as the state of the thermostat. Obviously, anything that is a property of the thermostat that can change and affects its behavior is part of its state in a broad sense. However, you may have a much narrower sense, that considers only some attributes to define the state, e.g., operation mode ("heat", "cool") for example.

I was trying to follow a pattern where folks seem to define such constants to achieve consistency between devices (I guess). I can take this out again, as it is just cosmetic.

MartinHjelmare replied:
I like having them as constants, but I don't think they should be named states as that is something specific in home assistant, and these do not qualify. We should think of something else.

The hold mode service has now been implemented in the climate component, so it does make sense to keep these constants (renamed) here in the component. Although I'm not sure that the hold mode service is a universal name for what it does.

Maybe I'm being picky about all these names, but I think that having good and accurate names help reading and explaining the code and understanding how home assistant works.

Duoxilian replied:
@MartinHjelmare I think you are correct in being picky, because consistent and meaningful naming is one of the best mechanisms of helping to keep the code maintainable. I appreciate your input.

nordlead2005 commented on homeassistant/components/climate/ecobee.py:
```diff
@@ -382,13 +383,3 @@ def hold_preference(self):
         else:
             return 'nextTransition'
-
-    # Sleep mode isn't used in UI yet:
```
FYI, Sleep mode can be a hold. Any climate (even ones you define) in ecobee can be a hold. See https://github.com/home-assistant/home-assistant/pull/5590

Now, does that mean we should keep this code in here? probably not, since it isn't used. I personally think that the toggle switch for "home" or "away" mode is the wrong choice and all of these functions should go away. I think we should move to an input select which would automatically be populated with home, away, temp, and sleep (well, each thermostat would have to provide their own list), and be extendable to add smart1, through smartX as the user needs, and it would be even better if the user can map meaningful names to the smartX climates.

Duoxilian replied:
I think you are confusing hold with schedule, see more detail in my comment to #5590.

Duoxilian replied:
Not sure what you mean by "toggle switch" for home or away mode, assuming you are referring to hold_mode. Obviously you would have an input selector in your UI as you are dealing with a choice between a number of alternatives, see the screen shot from my UI...
![screenshot from 2017-01-29 08-11-46](https://cloud.githubusercontent.com/assets/24535446/22404641/b5eed7ba-e5fa-11e6-8c4c-ac2e91fe886a.png)

Duoxilian replied:
@nordlead2005 Actually, no climate can be a hold. A climate can be referenced in a temperature hold, which means that certain values for the desired state are taken from that climate. For example, the fan operation or the participating sensors might be taken from the climate, rather than separately defined.

Duoxilian commented:
@MartinHjelmare I looked into the list of "states" in climate, and here is the summary: 
- values for ```climate.current_operation```
- ```STATE_FAN_ONLY```, unused. Presumably some kind of mode where only the fan is operating.
- ```STATE_DRY```, unused. Presumably a state of humitity
- values for ```climate.hold_mode``` (subject of this discussion).

So it seems to me there are two options: (i) consider any attribute that affects the operation of the thermostat part of its state, and refer to these values as such or (ii) consider only ```operation_mode``` as the state of the thermostat and remove the unused items above and rename or remove the hold mode items.

robbiet480 commented:
Hey, sorry the CLA link was wrong, it's now been fixed. Please sign the CLA [here](https://home-assistant.io/developers/cla_sign_start/).

Duoxilian commented:
@MartinHjelmare I decided to build ontop of what nordlead2005 suggested, well with some changes. But a consequence is that the constants you pointed out above would not be needed.

I am trying to coordinate with him these changes so we don't check in incompatible changes that cause merge problems, as I have some additions to what he proposed in #5590.

Duoxilian committed:
Remove use of constants for hold mode. Will be made irrelevant
by a planned change by nordlead2005.

Duoxilian commented:
@MartinHjelmare: I have removed the constants mentioned above, so that nordlead2005's updates are easier.

However, I have not heard from him since the weekend. I have already build some further enhancements based on his changes and don't want to get too far out of sync with the master. Do you think it is appropriate to check in his proposal and continue from there or should I wait until he responds? He does not have an email in his profile.

MartinHjelmare commented:
I'd wait for him to respond at least until next weekend, since he has a PR open with the changes you want to build on.

But I think we can merge this PR here now like this, right?

Duoxilian commented:
Yes, correct. I'll just break it into separate steps. Easier to review anyway.
